### PR TITLE
Improve performance of min_by and max_by

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

See commits

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java [hunk 1]
```diff
@@ -288,8 +290,8 @@ public HllState initialState(RamAccounting ramAccounting, MemoryManager memoryMa
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = FieldData.toString(DocValues.getSortedSet(reader, columnName));
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:293]: `leafReaderContext`

Comment (matriv) [inline extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:293]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java [hunk 2]
```diff
@@ -263,8 +265,8 @@ public MutableObject initialState(RamAccounting ramAccounting, MemoryManager mem
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:268]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java [hunk 3]
```diff
@@ -300,8 +302,8 @@ public MutableObject initialState(RamAccounting ramAccounting, MemoryManager mem
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = FieldData.toString(DocValues.getSortedSet(reader, columnName));
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java:305]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java [hunk 4]
```diff
@@ -98,8 +99,8 @@ public MutableLong initialState(RamAccounting ramAccounting, MemoryManager memor
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java:102]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java [hunk 5]
```diff
@@ -139,8 +140,8 @@ public MutableDouble initialState(RamAccounting ramAccounting, MemoryManager mem
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java:143]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java [hunk 6]
```diff
@@ -180,8 +181,8 @@ public MutableFloat initialState(RamAccounting ramAccounting, MemoryManager memo
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MaximumAggregation.java:184]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java [hunk 7]
```diff
@@ -97,8 +98,8 @@ public MutableLong initialState(RamAccounting ramAccounting, MemoryManager memor
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java:101]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java [hunk 8]
```diff
@@ -138,8 +139,8 @@ public MutableDouble initialState(RamAccounting ramAccounting, MemoryManager mem
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java:142]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java [hunk 9]
```diff
@@ -178,8 +179,8 @@ public MutableFloat initialState(RamAccounting ramAccounting, MemoryManager memo
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/MinimumAggregation.java:182]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/templates/SortedNumericDocValueAggregator.java [hunk 10]
```diff
@@ -56,8 +56,8 @@ public T initialState(RamAccounting ramAccounting, MemoryManager memoryManager,
     }
 
     @Override
-    public void loadDocValues(LeafReader reader) throws IOException {
-        values = DocValues.getSortedNumeric(reader, columnName);
+    public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/templates/SortedNumericDocValueAggregator.java:59]:
```suggestion
    public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/templates/BinaryDocValueAggregator.java [hunk 11]
```diff
@@ -57,8 +57,8 @@ public T initialState(RamAccounting ramAccounting, MemoryManager memoryManager,
     }
 
     @Override
-    public void loadDocValues(LeafReader reader) throws IOException {
-        values = FieldData.toString(DocValues.getSortedSet(reader, columnName));
+    public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/templates/BinaryDocValueAggregator.java:60]:
```suggestion
    public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java [hunk 12]
```diff
@@ -224,8 +226,8 @@ public NumericAverageState initialState(RamAccounting ramAccounting, MemoryManag
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java:229]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java [hunk 13]
```diff
@@ -264,8 +266,8 @@ public NumericAverageState initialState(RamAccounting ramAccounting, MemoryManag
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java:269]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java [hunk 14]
```diff
@@ -308,8 +310,8 @@ public NumericAverageState initialState(RamAccounting ramAccounting, MemoryManag
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java:313]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java [hunk 15]
```diff
@@ -326,8 +328,8 @@ public MutableFloat initialState(RamAccounting ramAccounting, MemoryManager memo
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java:331]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java [hunk 16]
```diff
@@ -288,8 +290,8 @@ public MutableDouble initialState(RamAccounting ramAccounting, MemoryManager mem
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java:293]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java [hunk 17]
```diff
@@ -254,8 +256,8 @@ public MutableLong initialState(RamAccounting ramAccounting, MemoryManager memor
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/SumAggregation.java:259]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java [hunk 18]
```diff
@@ -300,8 +302,8 @@ public BigDecimalValueWrapper initialState(RamAccounting ramAccounting, MemoryMa
         }
 
         @Override
-        public void loadDocValues(LeafReader reader) throws IOException {
-            values = DocValues.getSortedNumeric(reader, columnName);
+        public void loadDocValues(LeafReaderContext reader) throws IOException {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java:305]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java [hunk 19]
```diff
@@ -253,8 +255,8 @@ public BigDecimalValueWrapper initialState(RamAccounting ramAccounting, MemoryMa
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java:258]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java [hunk 20]
```diff
@@ -209,8 +211,8 @@ public OverflowAwareMutableLong initialState(RamAccounting ramAccounting, Memory
...
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/NumericSumAggregation.java:214]:
```suggestion
        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java [hunk 21]
```diff
@@ -105,6 +154,35 @@ public CmpByAggregation(int cmpResult, Signature signature, BoundSignature bound
         }
     }
 
+    @Override
+    public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver referenceResolver,
+                                                       List<Reference> aggregationReferences,
+                                                       DocTableInfo table,
+                                                       List<Literal<?>> optionalParams) {
+        Reference searchRef = aggregationReferences.get(1);
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:162]: Could you please add a comment why this is `1` and line 174 is `0`?

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java [hunk 22]
```diff
@@ -171,4 +250,82 @@ public Object terminatePartial(RamAccounting ramAccounting, CompareBy state) {
...
+
+        @Override
+        public void loadDocValues(LeafReaderContext leafReaderContext) throws IOException {
+            this.leafReaderContext = leafReaderContext;
+            values = DocValues.getSortedNumeric(leafReaderContext.reader(), columnName);
+        }
+
+        @Override
+        public void apply(RamAccounting ramAccounting, int doc, CmpByLongState state) throws IOException {
+            if (values.advanceExact(doc) && values.docValueCount() == 1) {
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:303]: `values.docValueCount() == 1` means not an array, right?
What happens if there is an array, do we guard this in analysis already?

**Review by matriv — COMMENTED** [review]
Great stuff!!
 - I left a bunch of the same comment regarding `LeafContextReader` arg name.
 - Left a couple of comments/questions on the actual impl.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java [hunk 23]
```diff
@@ -171,4 +250,82 @@ public Object terminatePartial(RamAccounting ramAccounting, CompareBy state) {
...
+                long value = values.nextValue();
+                if (value * mod > state.cmpValue * mod) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:305]: Long.MIN_VAL and MAX_VAL differ to 1 (by abs) which I think can lead to some overflow issues. Found at least one failing scenario, tested it locally

```
Signature signature = Signature.aggregate(
            "min_by",
            DataTypes.STRING.getTypeSignature(),
            DataTypes.LONG.getTypeSignature(),
            DataTypes.STRING.getTypeSignature()
        );
        Object result = executeAggregation(
            signature,
            new Object[][] {
                new Object[] { "a", 1l },
                new Object[] { "b", 2l },
                new Object[] { "c", Long.MIN_VALUE},
            },
            List.of()
        );
        assertThat(result).isEqualTo("c");
```
in debugger `partialResultWithoutDocValues` returns correct result "c"

not sure about other combinations - Long.MAX_VAL + (min/max by).   Long.Min_VALUE + `max_by` seems to be working.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java (hunk 21 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:162]: Doesn't the name `searchRef` and `resultExpression` give enough indication that it refers to the search and result parameter of the function?

File: extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java (hunk 1 above)
Comment (matriv) [reply extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:293]: As you like, just felt a bit weird to have `reader.reader()`, but it's really minor, no need to change.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java (hunk 23 above)
Comment (matriv) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java]: nice catch @BaurzhanSakhariev!

**Review by BaurzhanSakhariev — APPROVED** [review]
👍

**Review by matriv — APPROVED** [review]

**Pull request merged**